ddtrace/tracer/option.go

```diff
@@ -135,6 +135,75 @@ func defaults(c *config) {
 	}
 }
 
+// fallbackToTags checks if env, service, and version are set. If not,
+// it tries to load their values from globalTags.
+func (c *config) fallbackToTags() {
+	if c.env == "" {
+		if v, ok := c.globalTags["env"]; ok {
+			if e, ok := v.(string); ok {
+				c.env = e
+			}
+		}
+	}
+	if c.serviceName == "" {
+		if v, ok := c.globalTags["service"]; ok {
+			if s, ok := v.(string); ok {
+				c.serviceName = s
+				globalconfig.SetServiceName(s)
+			}
+		}
+	}
+	if c.version == "" {
+		if v, ok := c.globalTags["version"]; ok {
+			if ver, ok := v.(string); ok {
+				c.version = ver
+			}
+		}
+	}
+}
+
+// defaults ensures that certain fields in the tracer are set to their default
+// values, if a value has not been configured.
+func (c *config) defaults() {
+	if c.serviceName == "" {
+		c.serviceName = filepath.Base(os.Args[0])
+	}
+	if c.transport == nil {
+		c.transport = newTransport(c.agentAddr, c.httpClient)
+	}
+	if c.propagator == nil {
+		c.propagator = NewPropagator(nil)
+	}
+	if c.logger != nil {
+		log.UseLogger(c.logger)
+	}
+	if c.debug {
+		log.SetLevel(log.LevelDebug)
+	}
+	if c.statsd == nil {
+		client, err := statsd.New(c.dogstatsdAddr, statsd.WithMaxMessagesPerPayload(40), statsd.WithTags(statsTags(c)))
+		if err != nil {
+			log.Warn("Runtime and health metrics disabled: %v", err)
+			c.statsd = &statsd.NoOpClient{}
+		} else {
+			c.statsd = client
+		}
+	}
+}
+
+func newConfig(opts ...StartOption) *config {
+	c := new(config)
+	c.sampler = NewAllSampler()
+	c.agentAddr = defaultAddress
+	c.loadEnv()
+	for _, fn := range opts {
+		fn(c)
+	}
+	c.fallbackToTags()
```

Comment:
The name of this function implies that something happened and we need to fallback to something else. The code however indicates no problem. I think that in this case a big long function might be better than a split.

Reply:
The "fallback" is just that if the service, env, version are unset, we "fall back" to picking them up from the env.

We can definitely change the name to something better.

Reply:
I personally feel it saves you a bit of time as a reader to have one big long function. This way there's additional "side" functions which aren't really there to be re-used but just to presumably make the code easier to read. But they code could be easier to read by reading Go instead. No need to memorise new names nor check where everything is called for to make sense of things. No branching out, just a linear function.

> The "fallback" is just that if the service, env, version are unset, we "fall back" to picking them up from the env.

You only "fallback" if they are unset. So the function name does not necessarily mean that this will happen, thus the confusion.

> We can definitely change the name to something better.

I'm not sure if I know any better names than suggested. IMHO one big function will be clearer. If not, maybe `postConfig` or something like that?

Reply:
One big function works for me.